File: classes/what_compilers_can_and_cannot_do_9/c/nbody.c
//
// Created by Elvis Sabanovic on 30/12/2024.
//
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// LAW OF GRAVITATION:
// F_21 = (G m1m2/|r_12|^2) unit(r_21)

typedef struct vect_t {
  double x, y;
} vec_t;

static vec_t add(vec_t a, vec_t b) {
  vec_t sum = { a.x + b.x, a.y + b.y };
  return sum;
}

static vec_t scale(vec_t v, double a) {
  vec_t scaled = { v.x * a, v.y * a };
  return scaled;
}

static double vec_length2(vec_t v) {
  return v.x * v.x + v.y * v.y;
}

typedef struct body_t {
  vec_t position;
  vec_t velocity;
  vec_t force;
  double mass;
} body_t;

void update_position(int nbodies, body_t *bodies, double time_quantum) {
  for (int i = 0; i < nbodies; i++) {
    // compute new veolicty for each body
    vec_t new_velocity = scale(bodies[i].force, time_quantum / bodies[i].mass);

    // update position of each body based on average based on it's old and new velocity
    bodies[i].position = add(bodies[i].position, scale(add(bodies[i].velocity, new_velocity), time_quantum / 2.0));

    // set new velocity of bodies
    bodies[i].velocity = new_velocity;
  }
}

//void calculate_forces(int nbodies, body_t *bodies) {
//  for (int i = 0; i < nbodies; i++) {
//    for (int j = 0; j < nbodies; j++) {
//      if (i == j) continue;
//
//      add_force(&bodies[i], calculate_force(&bodies[i], &bodies[j]));
//    }
//  }
//}

//void calculate_forces(int nbodies, body_t *bodies) {
//  for (int i = 0; i < nbodies; i++) {
//    This change computes address on each j loop
//    we can hoist it to compute only when i ticks and re-use it in j loop
//    body_t *bi = &bodies[i];
//    for (int j = 0; j < nbodies; j++) {
//      if (i == j) continue;
//
//      add_force(body_t, calculate_force(body_t, &bodies[j]));
//    }
//  }
//}

void simulate(body_t *bodies, int nbodies, int nsteps, double time_quantum) {
  for (int i = 0; i < nbodies; i++) {
//    calculate_forces(nbodies, bodies);
    update_position(nbodies, bodies, time_quantum);
  }
}

int main() {
  srand(time(NULL));
  const int nbodies = 3;
  const int nsteps = 100;
  const double time_quantum = 0.1;

  body_t bodies[nbodies];

  // Initialize random bodies
  for (int i = 0; i < nbodies; i++) {
    bodies[i] = (body_t){
      .position = {(double)rand()/RAND_MAX * 20 - 10, (double)rand()/RAND_MAX * 20 - 10},
      .velocity = {0, 0},
      .force = {0, 0},
      .mass = (double)rand()/RAND_MAX * 100
  };
  }

  simulate(bodies, nbodies, nsteps, time_quantum);

  for (int i = 0; i < nbodies; i++) {
    printf("Body %d: Position (%.2f, %.2f)\n", i, bodies[i].position.x, bodies[i].position.y);
  }

  return 0;
}

File: classes/what_compilers_can_and_cannot_do_9/src/nbody.rs
use rand::Rng;

#[derive(Debug, Clone, Copy)]
struct Vec2 {
    x: f64,
    y: f64,
}

impl Vec2 {
    fn add(&self, other: Vec2) -> Vec2 {
        Vec2 {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }

    fn scale(&self, scalar: f64) -> Vec2 {
        Vec2 {
            x: self.x * scalar,
            y: self.y * scalar,
        }
    }

    fn length_squared(&self) -> f64 {
        self.x * self.x + self.y * self.y
    }
}

#[derive(Debug)]
struct Body {
    position: Vec2,
    velocity: Vec2,
    force: Vec2,
    mass: f64,
}

impl Default for Body {
    fn default() -> Self {
        let mut rng = rand::thread_rng();

        Self {
            position: Vec2 {
                x: rng.gen_range(-10.0..10.0),
                y: rng.gen_range(-10.0..10.0),
            },
            velocity: Vec2 { x: 0.0, y: 0.0 },
            force: Vec2 { x: 0.0, y: 0.0 },
            mass: rng.gen_range(1.0..100.0),
        }
    }
}

#[no_mangle]
fn update_position(bodies: &mut [Body], time_quantum: f64) {
    for body in bodies.iter_mut() {
        let new_velocity = body.force.scale(time_quantum / body.mass);
        body.position = body
            .position
            .add(body.velocity.add(new_velocity).scale(time_quantum / 2.0));
        body.velocity = new_velocity;
    }
}

fn simulate(bodies: &mut [Body], nsteps: i32, time_quantum: f64) {
    for _ in 0..nsteps {
        update_position(bodies, time_quantum);
    }
}

fn main() {
    let nbodies = 1000;
    let mut bodies: Vec<Body> = (0..nbodies).map(|_| Body::default()).collect();

    let time_quantum = 0.1;
    let nsteps = 10_000;

    simulate(&mut bodies, nsteps, time_quantum);

    for (i, body) in bodies.iter().enumerate() {
        println!(
            "Body {}: Position: ({:.2}, {:.2})",
            i, body.position.x, body.position.y
        );
    }
}
